File: tests/session_interface_test.py
#!/usr/bin/env python
# coding=UTF-8
# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4


from httpclient import Cookie, Session
import sys
import time

def test(X):
    if not X:
        raise Exception('Test failed')

def parse_info(resp):
    print 'Got info %s' % resp
    dic={}
    for pair in resp.split('\n'):
        dic[pair.split('=')[0]] = pair.split('=')[1]
    #print "Parsed as ", dic
    return dic

if len(sys.argv) < 3:
    print "Usage: (server|client|both) (fixed|renew|browser)"
    sys.exit(1)

def test_method_is(dic,method):
    v = dic['expiration']
    if method=='fixed':
        test(v=='0')
    elif method == 'renew':
        test(v=='1')
    else: # browser
        test(v=='2')

def general_tests(method):
    print "------------------- General Tests --------------------"
    session = Session()
    session.transmit('/new')
    cookie = session.received['sc']
    if method=='fixed' or method=='renew':
        test(cookie.max_age == '5')
    else:
        test(not cookie.max_age)
   
    test(cookie.domain == 'foo.bar')
    test(cookie.path == '/foo')

    time.sleep(2)

    session.transmit('/update')
    cookie = session.received['sc']
    
    if method=='fixed':
        test(int(cookie.max_age) < 5)
    elif method=='renew':
        test(int(cookie.max_age) == 5)
    
    test(session.transmit('/clear') == 'clear')
    test(session.received['sc'].max_age == '0')
    test(parse_info(session.transmit('/info'))['is_set_x']=='0')

    session.transmit('/new_short')
    if method!='browser':
        test(session.received['sc'].max_age == '1')
    time.sleep(2)
    test(session.transmit('/is_expired') == 'expired')

    print "new"
    session.transmit('/new')
    print "expose"
    session.transmit('/expose')
    test(session.received['sc_x'].value=='1')
    print "update"
    session.transmit('/update')
    test(session.received['sc_x'].value=='2')
    print "clear"
    session.transmit('/clear')
    test(session.received['sc_x'].max_age=='0')
    session.transmit('/new')
    test(parse_info(session.transmit('/info'))['is_exposed_x']=='0')
    session.transmit('/expose')
    test(session.received['sc_x'].value=='1')
    test(parse_info(session.transmit('/info'))['is_exposed_x']=='1')
    session.transmit('/unexpose')
    test(session.received['sc_x'].max_age=='0')
    test(parse_info(session.transmit('/info'))['is_exposed_x']=='0')
    test('sc' in session.state)
    session.transmit('/clear')
    test(session.received['sc'].max_age=='0')
    test(not 'sc_x' in session.received)

    session.transmit('/new')
    dic = parse_info(session.transmit('/info'))
    test(dic['is_set_x']=='1')
    test(dic['age']=='5')
    test_method_is(dic,method)
    session.transmit('/fixed')
    test(session.received['sc'].max_age=='5')
    test_method_is(parse_info(session.transmit('/info')),'fixed')
    time.sleep(2);
    test_method_is(parse_info(session.transmit('/info')),'fixed')
    test(not 'sc' in session.received)
    session.transmit('/renew')
    test(session.received['sc'].max_age=='5')
    test_method_is(parse_info(session.transmit('/info')),'renew')
    time.sleep(2);
    test_method_is(parse_info(session.transmit('/info')),'renew')
    test(session.received['sc'].max_age=='5')
    session.transmit('/browser')
    test(not session.received['sc'].max_age)
    test_method_is(parse_info(session.transmit('/info')),'browser')
    session.transmit('/clear')
    session.transmit('/new')
    test(parse_info(session.transmit('/info'))['age']=='5')
    session.transmit('/new_short')
    test(parse_info(session.transmit('/info'))['age']=='1')
    session.transmit('/clear')
    test(session.transmit('/api')=='ok')



def reset_session(x):
    print "------------------- Reset Session --------------------"
    session=Session()
    session.transmit('/new')
    prev_cookie = session.state['sc'].value
    session.transmit('/reset')
    new_cookie = session.state['sc'].value
    test(prev_cookie != new_cookie)
    test(prev_cookie[0]=='I')
    test(new_cookie[0]=='I')
    test(parse_info(session.transmit('/info'))['is_set_x']=='1')
def no_replay(x):
    session=Session()
    session.transmit('/new')
    old_state = session.state
    session.transmit('/clear')
    session.state = old_state
    test(parse_info(session.transmit('/info'))['is_set_x']=='0')
def force_server(x):
    print "------------------- Force Server --------------------"
    session=Session()
    session.transmit('/new')
    session.transmit('/update')
    test(session.state['sc'].value[0]=='C')
    test(parse_info(session.transmit('/info'))['on_server']=='0')
    session.transmit('/on_server')
    test(session.state['sc'].value[0]=='I')
    test(parse_info(session.transmit('/info'))['on_server']=='1')
    session.transmit('/update')
    test(session.state['sc'].value[0]=='I')
    test(parse_info(session.transmit('/info'))['on_server']=='1')
    session.transmit('/not_on_server')
    test(session.state['sc'].value[0]=='C')
    test(parse_info(session.transmit('/info'))['on_server']=='0')
    session.transmit('/update')
    test(session.state['sc'].value[0]=='C')
    test(parse_info(session.transmit('/info'))['on_server']=='0')
def size_tests(x):
    print "------------------- Force Server --------------------"
    session=Session()
    session.transmit("/huge")
    test(session.state['sc'].value[0]=='I')
    session.transmit("/small")
    test(session.state['sc'].value[0]=='C')
    session=Session()
    session.transmit("/new")
    test(session.state['sc'].value[0]=='C')
    session.transmit("/huge")
    test(session.state['sc'].value[0]=='I')
    session.transmit("/on_server")
    test(session.state['sc'].value[0]=='I')
    session.transmit("/not_on_server")
    test(session.state['sc'].value[0]=='I')
    session.transmit("/small")
    test(session.state['sc'].value[0]=='C')

method = sys.argv[2]


if sys.argv[1] == 'server':
    general_tests(method)
    reset_session(method)
    no_replay(method)
elif sys.argv[1] == 'client':
    general_tests(method)
elif sys.argv[1] == 'both':
    general_tests(method)
    size_tests(method)
    force_server(method)

File: tests/session_interface_test.cpp
///////////////////////////////////////////////////////////////////////////////
//                                                                             
//  Copyright (C) 2008-2012  Artyom Beilis (Tonkikh) <artyomtnk@yahoo.com>     
//                                                                             
//  See accompanying file COPYING.TXT file for licensing details.
//
///////////////////////////////////////////////////////////////////////////////
#include <cppcms/service.h>
#include <cppcms/application.h>
#include <cppcms/applications_pool.h>
#include <cppcms/http_request.h>
#include <cppcms/http_response.h>
#include <cppcms/http_context.h>
#include <cppcms/http_cookie.h>
#include <cppcms/session_interface.h>
#include <cppcms/serialization.h>
#include <cppcms/json.h>
#include <iostream>
#include "client.h"
#include "test.h"

struct mydata : public cppcms::serializable { 
	int x;
	int y;
	void serialize(cppcms::archive &a)
	{
		a & x & y;
	}
};

void test_set(cppcms::session_interface &interface,std::string const &values)
{
	std::set<std::string> keys = interface.key_set();
	std::string res;
	for(std::set<std::string>::const_iterator p=keys.begin();p!=keys.end();++p) {
		if(res.empty())
			res = *p;
		else
			res += "|" + *p;
	}
	if(res!=values)
		throw std::runtime_error("Key set is invalid:(" + res +")!=(" + values +")");
}

class adapter : public cppcms::session_interface_cookie_adapter {
public:
	adapter(cppcms::http::context &c) 
	{
		cookie_name = c.session().session_cookie_name();
		value = c.request().get("sid");
	}
	
	void set_cookie(cppcms::http::cookie const &updated_cookie)
	{
		if(updated_cookie.name() == cookie_name) {
			if((updated_cookie.max_age_defined() && updated_cookie.max_age() == 0)
			   || (updated_cookie.expires_defined() && updated_cookie.expires() + 1 < time(0)))
				value.clear();
			else
				value = updated_cookie.value();
		}
	}
	std::string get_session_cookie(std::string const &/*name*/)
	{
		return value;
	}
	std::set<std::string> get_cookie_names() { return std::set<std::string>(); }
	std::string value;
	std::string cookie_name;
};

class unit_test : public cppcms::application {
public:
	unit_test(cppcms::service &s) : cppcms::application(s)
	{
	}
	virtual void main(std::string u)
	{
		if(u=="/new") {
			session().set("x",1);
		}
		else if(u=="/update") {
			session().set("x",session().get<int>("x") + 1);
		}
		else if(u=="/clear") {
			session().clear();
			response().out() << "clear";
		}
		else if(u=="/new_short") {
			session().set("x",1);
			session().age(1);
		}
		else if(u=="/is_expired") {
			if(session().is_set("x"))
				response().out() << "not expired";
			else
				response().out() << "expired";
		}
		else if(u=="/expose") {
			session().expose("x");
		}
		else if(u=="/unexpose") {
			session().hide("x");
		}
		else if(u=="/fixed") {
			session().expiration(cppcms::session_interface::fixed);
		}
		else if(u=="/renew") {
			session().expiration(cppcms::session_interface::renew);
		}
		else if(u=="/browser") {
			session().expiration(cppcms::session_interface::browser);
		}
		else if(u=="/reset") {
			session().reset_session();
		}
		else if(u=="/on_server") {
			session().on_server(true);
		}
		else if(u=="/not_on_server") {
			session().on_server(false);
		}
		else if(u=="/huge") {
			std::string tmp="x";
			for(int i=0;i<10;i++)
				tmp=tmp+tmp;
			session().set("y",tmp);
		}
		else if(u=="/small") {
			session().erase("y");
			session().set("x","1");
		}
		else if(u=="/info") {
			bool is_set_x = session().is_set("x");
			bool is_exposed_x = session().is_exposed("x");
			int age = session().age();
			int expiration = session().expiration();
			bool on_server = session().on_server();
			response().out() 
				<< "is_set_x=" << is_set_x << "\n"
				<< "is_exposed_x=" << is_exposed_x << "\n"
				<< "age="<<age<<"\n"
				<< "expiration=" << expiration << "\n"
				<< "on_server=" << on_server ;
		}
		else if(u=="/api") {
			try {
				// Fix Me Later
				test_set(session(),"");
				session().set("x","10");
				test_set(session(),"x");
				TEST(session().get<std::string>("x")=="10");
				TEST(session().get<int>("x")==10);
				TEST(session().get<double>("x")==10.0);
				TEST(session().get("z","default")=="default");
				TEST(session().get("x","default")=="10");
				TEST(session()["x"]=="10");
				TEST(session()["z"]=="");
				test_set(session(),"x|z");
				session()["z"]="test";
				TEST(session()["z"]=="test");

				mydata a,b;
				a.x=10;
				a.y=20;
				session().store_data("tmp",a);
				session().fetch_data("tmp",b);
				test_set(session(),"tmp|x|z");
				TEST(b.x==10);
				TEST(b.y==20);


				response().out() << "ok";
			}
			catch(std::exception const &e) {
				std::cerr << "Failed " << e.what() << std::endl;
				response().out() << "not ok";
			}
		}
		else if(u=="/alt") {
			adapter a(context());
			session().set_cookie_adapter_and_reload(a);
			bool is_set=session().is_set("x");
			std::string val;
			if(is_set)
				val=session().get("x");
			if(request().get("x")!="")
				session().set("x",request().get("x"));
			response().out() << "is_set_x="<<is_set << '\n'
					 << "x='" << val << "'\n";
			response().out() << "sid=" << a.value;
		}
	}
};





int main(int argc,char **argv)
{
	try {
		cppcms::service srv(argc,argv);
		srv.applications_pool().mount( cppcms::create_pool<unit_test>());
		srv.after_fork(submitter(srv));
		srv.run();
	}
	catch(std::exception const &e) {
		std::cerr << e.what() << std::endl;
		return EXIT_FAILURE;
	}
	return run_ok ? EXIT_SUCCESS : EXIT_FAILURE;
}
